adapter/serviceControl/serviceController.go

```diff
@@ -0,0 +1,197 @@
+// Copyright 2017 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package serviceControl
+
+import (
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/pborman/uuid"
+	servicecontrol "google.golang.org/api/servicecontrol/v1"
+
+	"istio.io/mixer/adapter/serviceControl/config"
+	"istio.io/mixer/pkg/adapter"
+)
+
+type (
+	builder struct {
+		adapter.DefaultBuilder
+
+		client
+	}
+
+	aspect struct {
+		serviceName string
+		service     *servicecontrol.Service
+		logger      adapter.Logger
+	}
+)
+
+var (
+	name = "serviceControl"
+)
+
+// Register records the builders exposed by this adapter.
+func Register(r adapter.Registrar) {
+	r.RegisterMetricsBuilder(
+		&builder{adapter.NewDefaultBuilder(
+			name,
+			"Push metrics to GCP service controller",
+			new(config.Params),
+		), new(clientImpl)})
+
+	r.RegisterApplicationLogsBuilder(
+		&builder{adapter.NewDefaultBuilder(
+			name,
+			"Writes log entries to GCP service controller",
+			new(config.Params),
+		), new(clientImpl)})
+}
+
+func (*builder) ValidateConfig(c adapter.Config) (ce *adapter.ConfigErrors) {
+	return nil
+}
+
+func (b *builder) NewMetricsAspect(env adapter.Env, cfg adapter.Config, metrics map[string]*adapter.MetricDefinition) (adapter.MetricsAspect, error) {
+	return b.newAspect(env, cfg)
+}
+
+func (b *builder) NewApplicationLogsAspect(env adapter.Env, cfg adapter.Config) (adapter.ApplicationLogsAspect, error) {
+	return b.newAspect(env, cfg)
+}
+
+func (*builder) NewAccessLogsAspect(env adapter.Env, cfg adapter.Config) (adapter.AccessLogsAspect, error) {
+	return nil, errors.New("access logs are not supported")
+}
+
+func (b *builder) newAspect(env adapter.Env, cfg adapter.Config) (*aspect, error) {
+	params := cfg.(*config.Params)
+	ss, err := b.client.create(env.Logger())
+	if err != nil {
+		return nil, err
+	}
+	return &aspect{params.ServiceName, ss, env.Logger()}, nil
+}
+
+func (a *aspect) Record(values []adapter.Value) error {
+	rq := a.record(values, time.Now(), fmt.Sprintf("mixer-metric-report-id-%d", uuid.New()))
```

Comment:
`uuid.New()` returns a `string`, but your format string is using `%d`.

One thing we _might_ consider is using a `bytes.Buffer` or simple string concat (via `+`), instead of `fmt`. This can be done later if it proves to be an issue, but since this will be done over and over, it might be nice to just make it more efficient now.

```
op := bytes.New("mixer-metric-report-id-").WriteString(uuid.New())
rq := a.record(values, time.Now(), op.String())
```
or

```
rq := a.record(values, time.Now(), "mixer-metric-report-id-" + uuid.New())
```